cpp/src/arrow/compute/exec/filter_node.cc

```diff
@@ -127,21 +131,24 @@ class FilterNode : public ExecNode {
     StopProducing();
   }
 
-  void StopProducing() override { inputs_[0]->StopProducing(this); }
+  void StopProducing() override {
+    if (batch_count_.Cancel()) {
+      this->MarkFinished(/*request_stop=*/true);
+    }
+    inputs_[0]->StopProducing(this);
+  }
 
-  Future<> finished() override { return inputs_[0]->finished(); }
+  Future<> finished() override { return finished_; }
```

Comment:
Maybe this should be the default implementation in the base class now? (Though I guess for this PR that requires more refactoring.)

Reply:
(And in general, if we're codifying patterns like `AtomicCounter batch_count_`, I think more of these boilerplate methods could be implemented in the base class instead, once we're OK with the overall design in this PR.)